### PR TITLE
Issue-134: allow configurable and dynamic s3 path

### PR DESCRIPTION
**Pull request by arihantsurana**
This PR solves for https://github.com/snowplow/snowplow-s3-loader/issues/134 by enabling the use of `DateTimeFormat` patterns enclosed in {curly braces} as an optional s3 configuration - `directoryPattern`.

The configuration can be used to supply a static directory or use the emitter's connection start time to fill in dynamic values replacing time formatted chunks in braces. All values which do not conform the `DateTimeFormat` are left as is.

### CODE REVIEW

**Comment by snowplowcla** [comment]
@arihantsurana has signed the Software Grant and Corporate Contributor License Agreement

File: examples/config.hocon.sample [hunk 1]
```diff
@@ -86,6 +86,9 @@ streams {
 s3 {
   region = "{{s3Region}}"
   bucket = "{{s3bucket}}"
+  # optional directory structure to use while storing data on s3, you can place date format strings which will be
+  # replaced with date time values
```
Comment (BenFradet) [inline examples/config.hocon.sample]: I think we'd need an example

File: src/main/scala/com.snowplowanalytics.s3/loader/S3Emitter.scala [hunk 2]
```diff
@@ -153,14 +156,32 @@ class S3Emitter(
...
+            (str, converted.getOrElse(str))
+          }
+          )
+          .toList
+        val directory = replacements.foldLeft(directoryPattern.getOrElse("")) {
+          (dir, replacement) => dir.replace(replacement._1, replacement._2)
+        }
+        // ensure trailing slash in the directory only if a pattern was supplied
+        val filename = if (directory.endsWith("/") || directory.isEmpty) s"$directory${namedStream.filename}"
+        else s"$directory/${namedStream.filename}"
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.s3/loader/S3Emitter.scala]: can we extract all this into its own standalone function and add tests?

File: examples/config.hocon.sample (hunk 1 above)
Comment (arihantsurana) [reply examples/config.hocon.sample]: Added example in comments.

File: src/main/scala/com.snowplowanalytics.s3/loader/S3Emitter.scala (hunk 2 above)
Comment (arihantsurana) [reply src/main/scala/com.snowplowanalytics.s3/loader/S3Emitter.scala]: Refactored s3 path calculation to a separate method and added unit tests

**Comment by arihantsurana** [comment]
@BenFradet added a standalone function with unit tests

File: src/main/scala/com.snowplowanalytics.s3/loader/S3Emitter.scala [hunk 3]
```diff
@@ -21,6 +21,8 @@ package com.snowplowanalytics.s3.loader
 //Java
 import java.io.ByteArrayInputStream
 
+import scala.util.Try
```
[the remark this answers is not in the document]
Comment (arihantsurana) [reply src/main/scala/com.snowplowanalytics.s3/loader/S3Emitter.scala]: Done.

File: src/test/scala/com/snowplowanalytics/s3/loader/DynamicPathSpec.scala [hunk 4]
```diff
@@ -0,0 +1,38 @@
+package com.snowplowanalytics.s3.loader
```
[the remark this answers is not in the document]
Comment (arihantsurana) [reply src/test/scala/com/snowplowanalytics/s3/loader/DynamicPathSpec.scala]: Done.

File: src/main/scala/com.snowplowanalytics.s3/loader/DynamicPath.scala [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    // replace the pattern with actual date values
+    val detectBracesExpression = "\\{(.*?)\\}".r
+    val replacements = detectBracesExpression
+      .findAllMatchIn(directoryPattern)
+      .map(_.toString.trim)
+      .map(str => {
+        val converted = Try(DateTimeFormat.forPattern(str).withZone(DateTimeZone.UTC).print(decoratorDateTime))
+        (str, converted.getOrElse(str))
+      }
+      )
```
[the remark this answers is not in the document]
Comment (arihantsurana) [reply src/main/scala/com.snowplowanalytics.s3/loader/DynamicPath.scala]: Done.

File: src/main/scala/com.snowplowanalytics.s3/loader/DynamicPath.scala [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import org.joda.time.{DateTime, DateTimeZone}
+import org.joda.time.format.DateTimeFormat
+
+import scala.util.Try
+
+/**
+  * Object to handle S3 dynamic path generation
+  */
+object DynamicPath {
+  def decorateDirectoryWithTime(directoryPattern: String, fileName: String, decoratorDateTime: DateTime): String = {
```
[the remark this answers is not in the document]
Comment (arihantsurana) [reply src/main/scala/com.snowplowanalytics.s3/loader/DynamicPath.scala]: Done. and Done.

**Comment by christoph-buente** [comment]
Hi @BenFradet,

is this feature planned for the next release? We are have a lamda function in place to remedy the lack of it :)

Thank you!

**Comment by BenFradet** [comment]
hey @christoph-buente, yes it'll be in the next release :+1:

**Comment by christoph-buente** [comment]
Cool. I was checking the milestone field above, but it was not maked. Found the version tag on the commit though.

**Pull request merged**